### PR TITLE
DeepSpeed ZeRO Docs update

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Adds some more information about ZeRO Stage 3, but still missing a reasonable amount of info. Lots of additional stuff to come once we iron out of some the current kinks with the API.

Fixes #<issue_number>

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/advanced/multi_gpu.rst [hunk 1]
```diff
@@ -788,6 +788,167 @@ For even more speed benefit, DeepSpeed offers an optimized CPU version of ADAM c
...
 
+DeepSpeed ZeRO Stage 3
+""""""""""""""""""""""
+
+DeepSpeed ZeRO Stage 3 shards the optimizer states, gradients and the model parameters (also optionally activations). Sharding model parameters and activations comes with an increase in distributed communication, however allows you to scale your models massively from one GPU to multiple GPUs.
+**The DeepSpeed team report the ability to fine-tune models with over 40B parameters on a single GPU and over 2 Trillion parameters on 512 GPUs.** For more information we suggest checking the `DeepSpeed ZeRO-3 Offload documentation <https://www.deepspeed.ai/news/2021/03/07/zero3-offload.html>`__.
+
+We've ran benchmarks and give a simple example of how all these features in Lightning, which you can see at `minGPT <https://github.com/SeanNaren/minGPT/tree/stage3>`_.
+
+Currently this functionality is only available on master and will be included in our next 1.3 Release Candidate and 1.3 release.
```
Comment (kaushikb11) [inline docs/source/advanced/multi_gpu.rst:799]:
```suggestion
Currently this functionality is only available on master and will be included in our next 1.3 Release Candidate and 1.3 release.
```
Will this be necessary? As we wouldn't be including this PR in the patch release.

Comment (SeanNaren) [reply docs/source/advanced/multi_gpu.rst:799]: Not sure if it is, can remove it people prefer!

**Review by kaushikb11 — APPROVED** [review]
Awesome!

File: docs/source/advanced/multi_gpu.rst (hunk 1 above)
Comment (Borda) [reply docs/source/advanced/multi_gpu.rst:799]: I would drop it...

File: docs/source/advanced/multi_gpu.rst [hunk 2]
```diff
@@ -788,6 +788,167 @@ For even more speed benefit, DeepSpeed offers an optimized CPU version of ADAM c
...
+
+.. code-block:: python
+
+    pip install git+https://github.com/PyTorchLightning/pytorch-lightning.git
```
Comment (Borda) [inline docs/source/advanced/multi_gpu.rst:803]:
```suggestion
    pip install https://github.com/PyTorchLightning/pytorch-lightning/archive/refs/heads/master.zip
```
git is already heavy

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Looks great !

File: docs/source/advanced/multi_gpu.rst [hunk 3]
```diff
@@ -788,6 +788,169 @@ For even more speed benefit, DeepSpeed offers an optimized CPU version of ADAM c
     trainer = Trainer(gpus=4, plugins=DeepSpeedPlugin(cpu_offload=True), precision=16)
     trainer.fit(model)
 
+DeepSpeed ZeRO Stage 3
+""""""""""""""""""""""
+
+DeepSpeed ZeRO Stage 3 shards the optimizer states, gradients and the model parameters (also optionally activations). Sharding model parameters and activations comes with an increase in distributed communication, however allows you to scale your models massively from one GPU to multiple GPUs.
+**The DeepSpeed team report the ability to fine-tune models with over 40B parameters on a single GPU and over 2 Trillion parameters on 512 GPUs.** For more information we suggest checking the `DeepSpeed ZeRO-3 Offload documentation <https://www.deepspeed.ai/news/2021/03/07/zero3-offload.html>`__.
+
+We've ran benchmarks and give a simple example of how all these features in Lightning, which you can see at `minGPT <https://github.com/SeanNaren/minGPT/tree/stage3>`_.
```
Comment (awaelchli) [inline docs/source/advanced/multi_gpu.rst:797]:
```suggestion
We've ran benchmarks and give a simple example of all these features in Lightning, which you can see at `minGPT <https://github.com/SeanNaren/minGPT/tree/stage3>`_.
```
of all these features
of how all these features work

File: docs/source/advanced/multi_gpu.rst [hunk 4]
```diff
@@ -788,6 +788,169 @@ For even more speed benefit, DeepSpeed offers an optimized CPU version of ADAM c
...
+2. Use CPU Offloading to offload weights to CPU, plus have a reasonable amount of CPU RAM to offload onto
+3. Use DeepSpeed Activation Checkpointing to shard activations
+
+Below we describe how to enable all of these to see benefit. **With all these improvements we reached 45 Billion parameters training a GPT model on 8 GPUs with ~1TB of CPU RAM available**.
+
+Also please have a look at our :ref:`deepspeed-zero-stage-3-tips` which contains a lot of helpful information when configuring your own models.
+
+.. note::
+    Currently we only support non-elastic checkpointing. This means saving the model across GPUs will save shards of the model on all processes, which will then require the same amount of GPUS to load.
+    This additionally means for inference you must use the ``Trainer.test` or ``Trainer.predict`` functionality as described below, to ensure we set up the distributed environment correctly.
```
Comment (awaelchli) [inline docs/source/advanced/multi_gpu.rst:818]:
```suggestion
    This additionally means for inference you must use the ``Trainer.test`` or ``Trainer.predict`` functionality as described below, to ensure we set up the distributed environment correctly.
```

Comment (awaelchli) [reply docs/source/advanced/multi_gpu.rst:818]: unbalanced ticks

File: docs/source/advanced/multi_gpu.rst [hunk 5]
```diff
@@ -788,6 +788,169 @@ For even more speed benefit, DeepSpeed offers an optimized CPU version of ADAM c
...
+
+When instantiating really large models, it is sometimes necessary to shard the model layers instantly.
+
+This is the case if layers may not fit on one single machines CPU or GPU memory, but would fit once sharded across multiple machines.
+We expose a hook that layers initialized within the hook will be sharded instantly on a per layer basis, allowing you to instantly shard models.
+
+This reduces the time taken to initialize very large models, as well as ensure we do not run out of memory when instantiating larger models. For more information you can refer to the DeepSpeed docs for `Constructing Massive Models <https://deepspeed.readthedocs.io/en/latest/zero3.html>`_.
+
+.. note::
+    When using ``configure_sharded_model`` hook to shard models, note that ``LightningModule.load_from_checkpoint`` for loading saved checkpoints may not work. If you've trained on one GPU, you can manually instantiate the model and call the hook,
```
Comment (awaelchli) [inline docs/source/advanced/multi_gpu.rst:852]:
```suggestion
    When using the ``configure_sharded_model`` hook to shard models, note that ``LightningModule.load_from_checkpoint`` for loading saved checkpoints may not work. If you've trained on one GPU, you can manually instantiate the model and call the hook,
```

File: docs/source/advanced/multi_gpu.rst [hunk 6]
```diff
@@ -788,6 +788,169 @@ For even more speed benefit, DeepSpeed offers an optimized CPU version of ADAM c
...
+
+Below we describe how to enable all of these to see benefit. **With all these improvements we reached 45 Billion parameters training a GPT model on 8 GPUs with ~1TB of CPU RAM available**.
+
+Also please have a look at our :ref:`deepspeed-zero-stage-3-tips` which contains a lot of helpful information when configuring your own models.
+
+.. note::
+    Currently we only support non-elastic checkpointing. This means saving the model across GPUs will save shards of the model on all processes, which will then require the same amount of GPUS to load.
+    This additionally means for inference you must use the ``Trainer.test` or ``Trainer.predict`` functionality as described below, to ensure we set up the distributed environment correctly.
+
+    This limitation is actively being worked on and will be resolved in the near future.
```
Comment (awaelchli) [inline docs/source/advanced/multi_gpu.rst:820]: totally naive idea:

we save the files like this:

checkpoint_name_shard0.ckpt
checkpoint_name_shard1.ckpt
checkpoint_name_shard2.ckpt
...

if user loads a checkpoint that ends with _shard_x and we detect other shard checkpoints in the same folder, then we do a combined loading?

File: docs/source/advanced/multi_gpu.rst [hunk 7]
```diff
@@ -788,6 +788,169 @@ For even more speed benefit, DeepSpeed offers an optimized CPU version of ADAM c
...
+    )
+    trainer.fit(model)
+
+
+.. _deepspeed-zero-stage-3-tips:
+
+DeepSpeed ZeRO Stage 3 Tips
+"""""""""""""""""""""""""""
+
+Here are some helpful information when setting up DeepSpeed ZeRO Stage 3 with Lightning.
```
Comment (awaelchli) [inline docs/source/advanced/multi_gpu.rst:948]: is information plural here?

**Review by awaelchli — APPROVED** [review]

**Pull request merged**